src/base/cub_ddl_log.c

```diff
@@ -0,0 +1,1206 @@
+    {
+      retval =
+	snprintf (filename_buf, buf_size, "%s/%s/%s_%s_ddl.log", env_root, DDL_LOG_PATH, cub_get_app_name (app_name),
+		  ddl_audit_handle->db_name);
+    }
+  else
+    {
+      filename_buf[0] = '\0';
+      return -1;
+    }
+
+  if (retval < 0)
+    {
+      assert (false);
+      filename_buf[0] = '\0';
+    }
+
+  return retval;
+}
+
+static FILE *
+cub_ddl_log_open (T_APP_NAME app_name)
+{
+  FILE *fp = NULL;
+  char *tpath = NULL;
+  int len;
+
+  fprintf (stderr, "=========== cub_ddl_log_open : %s \n", cub_get_app_name (app_name));
+
+  len = cub_make_ddl_log_filename (ddl_audit_handle->log_filepath, PATH_MAX, app_name);
+
+  if (ddl_audit_handle->log_filepath[0] == '\0' || len < 0)
+    {
+      goto file_error;
+    }
+
+  if (cub_create_dir_log (ddl_audit_handle->log_filepath) < 0)
+    {
+      goto file_error;
+    }
+
+  /* note: in "a+" mode, output is always appended */
+  fp = cub_fopen_and_lock (ddl_audit_handle->log_filepath, "a+");
+
+  return fp;
+
+file_error:
+  return NULL;
+}
+
+void
+cub_ddl_log_write_end ()
+{
+  if (ddl_audit_handle == NULL)
+    {
+      return;
+    }
+
+  if (prm_get_bool_value (PRM_ID_DDL_AUDIT_LOG) == FALSE)
+    {
+      goto ddl_log_free;
+    }
+
+  if (ddl_audit_handle->app_name == APP_NAME_LOADDB)
+    {
+      if (ddl_audit_handle->loaddb_file_type != LOADDB_FILE_TYPE_SCHEMA
+	  && ddl_audit_handle->loaddb_file_type != LOADDB_FILE_TYPE_INDEX
+	  && ddl_audit_handle->loaddb_file_type != LOADDB_FILE_TYPE_TRIGGER)
+	{
+	  goto ddl_log_free;
+	}
+    }
+  else
+    {
+      if (cub_is_ddl_type (ddl_audit_handle->stmt_type) == FALSE)
+	{
+	  goto ddl_log_free;
+	}
+    }
+  if (ddl_audit_handle->execute_type & DDL_LOG_RUN_EXECUTE_FUNC)
+    {
+      cub_ddl_log_write ();
+    }
+
+ddl_log_free:
+  cub_ddl_log_free (FALSE);
+}
+
+void
+cub_ddl_log_write ()
+{
+  FILE *fp = NULL;
+  char buf[DDL_LOG_BUFFER_SIZE] = { 0 };
+  char dest_path[PATH_MAX] = { 0 };
+  int len = 0;
+  int ret = 0;
+
+  if (ddl_audit_handle == NULL)
+    {
+      return;
+    }
+
+  if (prm_get_bool_value (PRM_ID_DDL_AUDIT_LOG) == FALSE)
+    {
+      return;
+    }
+
+  if (ddl_audit_handle->app_name == APP_NAME_LOADDB)
+    {
+      if (ddl_audit_handle->loaddb_file_type != LOADDB_FILE_TYPE_SCHEMA
+	  && ddl_audit_handle->loaddb_file_type != LOADDB_FILE_TYPE_INDEX
+	  && ddl_audit_handle->loaddb_file_type != LOADDB_FILE_TYPE_TRIGGER)
+	{
+	  return;
+	}
+    }
+  else
+    {
+      if (cub_is_ddl_type (ddl_audit_handle->stmt_type) == FALSE)
+	{
+	  return;
+	}
+    }
+
+  fp = cub_ddl_log_open (ddl_audit_handle->app_name);
+
+  if (fp != NULL)
+    {
+      if (ddl_audit_handle->app_name == APP_NAME_LOADDB)
+	{
+	  if (cub_make_schema_file_name (ddl_audit_handle->file_name, dest_path, PATH_MAX) < 0)
+	    {
+	      goto write_error;
+	    }
+	}
+
+      len = cub_create_log_mgs (buf);
+      if (len < 0)
+	{
+	  goto write_error;
+	}
+
+      if (fwrite (buf, 1, len, fp) != len)
```

Comment:
What about
if (fwrite (buf, sizeof(char), len, fp) != len)

Reply:
Thank you, I will change it to the #define statement.

Reply:
There are four more **fwrite**s of this style. Hope to fix it too.

Reply:
Thank you, i fixed it.